Implement Choice restriction support and fix some related issues

ain-is opened the pull request:
Hello!

  I have submitted Issue #20 few days ago. This is fix for it.

 Changes:
1. Implement Choice restriction support. Entries and values for choice restriction are used from application manifest restrictions definition.
2. Fix issue with incorrect String restriction type inside Bundle. If String restriction has null default value it was displayed as Boolean in UI inside Bundle.
3. Sort restrictions in alphabetic order inside Bundle.
4. When user creates new Bundle inside BundleArray restriction, init it with same restrictions as first existing bundle in same array instead of making empty Bundle.

ming030890 commented on app/src/main/java/com/afwsamples/testdpc/common/RestrictionManagerCompat.java:
```diff
@@ -31,6 +31,12 @@
     private static final String TAG = "RestrictionManager";
 
+    // These keys are used inside TesDPC only. We never save them to DevicePolicyManager
+    public static final String CHOICE_SELECTED_VALUE= "testdpc_arg_choice_selected_value";
```
pls insert a space before =

ming030890 commented:
Thanks, really appreciate it!
We will take a look and get back to you. :)

ming030890 commented on app/src/main/java/com/afwsamples/testdpc/common/RestrictionManagerCompat.java:
```diff
@@ -31,6 +31,12 @@
 public class RestrictionManagerCompat {
```
Modifying RestrictionManagerCompat to store some extra information for the UI sounds a bit hacky to me.
Also, RestrictionManagerCompat is ported from android.content.RestrictionsManager, so keeping it in sync with the framework make us easier to maintain.

IMO, the root cause of the bug is because we are mixing value(Bundle) and format(RestrictionEntry). They are actually two different things, but we keep converting between them!
How about having the fragment takes both value(Bundle) and format(RestrictionEntry) as input and render accordingly? With this, we can remove all the conversion code too.
It may requires a big change, so feel free to delegate to us if you do not have time. That said, we may not have it fixed in the near future as we are focusing on other things right now.

Feel free to chime in if you have any idea to get it fixed in a simple and clean way :)

Thanks!

ming030890 replied:
@AndrewScull  + Andrew to the discussion ☺️ 

Hmm, speaking of solution 1, I am not quite understand why putting RestrictionEnrty into the fragment argument bundle requires conversion to bundle. RestrictionEnrty is parcelable, and should be ok to put in the arguments bundle straight away?

```
        arguments.putBundle(ARG_INITIAL_VALUE, bundle);
        arguments.putparcelableArray(ARG_RESTRICTIONS, restrictions);

```

Thanks!

ain-is replied:
Yes, you are right. I missed, that RestrictionEntry is Parcelable. In this case solution 1 looks good. 
I can implement it.

ming030890 commented on app/src/main/java/com/afwsamples/testdpc/common/RestrictionManagerCompat.java:
```diff
@@ -47,36 +53,54 @@
+     * @param saveChoiceDataForDialog if true, then Choice restriction will be converted into Bundle with entries and values
      */
-    public static Bundle convertRestrictionsToBundle(List<RestrictionEntry> entries) {
+    public static Bundle convertRestrictionsToBundle(List<RestrictionEntry> entries, boolean saveChoiceDataForDialog) {
```
Please split the line that exceed the max char limit
Thanks

ain-is replied:
ok

ain-is commented:
I have added RestrictionEntry as parameter do implement approach discussed above.
KeyValuePairDialogFragment is used in SetTrustAgentConfigFragment also. By this reason I have modified it to support both RestrictionEntry and Bundle as parameters.

ming030890 commented:
Again, thanks for the effort.   :)
I will try to review it within this week.

ming030890 commented on app/src/main/java/com/afwsamples/testdpc/common/keyvaluepair/KeyValueBundleArrayFragment.java:
```diff
@@ -87,12 +95,37 @@ public static KeyValueBundleArrayFragment newInstance(String key, Bundle[] bundl
+        mBundleArrayRestrictionEntry = (RestrictionEntry) getArguments().getParcelable(
+                ARG_RESTRICTION_ENTRY);
+
+        if (mBundleArrayRestrictionEntry != null) {
```
As I said in previous comment, we have this kind of if else clause all over the codes now.

ain-is commented on app/src/main/java/com/afwsamples/testdpc/common/keyvaluepair/KeyValueBundleArrayFragment.java:
```diff
@@ -40,43 +41,50 @@
      * The current bundle list.
      */
-    private List<Bundle> mBundleList;
+    private ArrayList<Object> mBundleList;
```
Thanks for review.
Regarding ArrayList - yes, it can be changed to List.
Regarding Bundle and RestrictionEntry, looks like we really had some misunderstanding. In your old comment I see following: "How about having the fragment takes both value(Bundle) and format(RestrictionEntry) as input and render accordingly? **With this, we can remove all the conversion code too.**"
 But we cannot remove conversion code, if we will pass both Bundle and RestrictionEntry list as parameters for app restrictions. We will need to convert manifest restrictions into bundle, at least when application has no any restrictions values saved in DevicePolicyManager and manifest restrictions only are available.
 Also, passing same data twice (as Bundle list and as RestrictionEnty list) looks strange.
 One more point - with using Bundle list it will be difficult to show restrictions in same order as defined by application. But this is important, because in case of complex restrictions, application developer can define them in some specific order. For example, if restrictions defines some account settings, then user name probably will be first restriction to be dsplayed to user. But in case of alphabetic sorting it will be near the end of the restrictions list.
  Yes, there is some "if - else" code to handle both Bundle and RestrictionEntry, but it was added to support Bundles for SetTrustAgentConfigFragment only. For application restrictions it is much more convenient to use RestrictionEntry list and convert it into Bundle for saving into DevicePolicyManager only.

ming030890 replied:
> Also, passing same data twice (as Bundle list and as RestrictionEnty list) looks strange.

I see it in another way. The Bundle object represents the data, while RestrictionEntry represents the schema of the data. So, we are not passing the same thing twice, but passing both the data schema and the data itself. The UI should enforce the schema, and do not allow user to violate it (e.g: put
a string to a int field). After users finish updating the data (bundle), we save the bundle straight away without conversion. In the ideal world, trust agent config is just a special case that the schema have no restriction on the data at all (empty list of RestrictionEntry), same as the case that app have no manifest restriction declared.

Thanks

ain-is replied:
I understand your point, but I don't want to rework this code one more time. :)
Please, reject my pull requet and implement this yourself.

Thanks.

ming030890 replied:
Sure and no problem at all.
And I think you unblocked yourself anyway :)

abioteau commented:
Hi,

It works well with complex restrictions schema (BundleArray -> Bundle -> Choice ...).

Do you know, when it will be merged into master?

svc518 commented:
Is there an ETA on this merge?

fotoradar69 commented:
Jgt

abioteau commented:
@AndrewScull Please, could you review and merge this very old PR?

jscott1989 commented:
This is a very old PR.

I don't really have context on it - if it's still relevant can you fix the merge conflict and re-open this PR then I'll do a full review

ain-is commented:
Hi, 
This PR contains fix for issue https://github.com/googlesamples/android-testdpc/issues/20
Issue is still not resolved.
Unfortunately, I have no time to merge conflict or to do anythig else with this change. So, it is up to you, what to do with this.
